### ISSUE TITLE
[only_throw_errors] prevents explicit control flow

### ISSUE THREAD

**Issue by modulovalue** [dart-lang/linter#4053]
Consider the `only_throw_errors` lint rule and the following example:

```dart
//               vvvvv Only throw instances of classes extending either Exception or Error.
void aa() {throw foo();}
Never foo() {throw Exception();}
```

In this example only_throw_errors kicks in when a value of type Never is being thrown.

I think this goes against the general unspoken convention that explicit control flow should be preferred in Dart (_as opposed to Rust where among others, allowing the last expression in a list of statements to be returned implicitly is fine_)

I think `only_throw_errors` should not report on expressions that are known to be of type Never to allow users that would like to prefer explicit control flow (_i.e. to throw values of type Never instead of implicitly letting it jump_) to do that.

Related (in the sense that this also prefers implicit control flow): https://github.com/dart-lang/linter/issues/2352

**Comment by srawlins** [comment]
I don't think there is any practice to throw a value of type `Never`. The standard practice would be to call `foo()`. For example, to fail a test from package:test, you call `fail()`, but don't throw it.

**Comment by modulovalue** [comment]
> I don't think there is any practice to throw a value of type Never

Yes, I agree.

I meant to imply that Dart lacks implicit control flow support such as implicitly returning the last expression in a list of statements (e.g. as in Rust), so users are less likely to expect implicit control flow in Dart and implicit control flow may be perceived as surprising.

---

Since Never is the bottom type, wouldn't a value of type `Never` by definition be a subtype of Exception and Error? Wouldn't this mean that this is a false positive?

**Comment by lrhn** [comment]
Strictly speaking, `Never` is a subtype of both `Error` and `Exception`, but it doesn't *implement* `Error` or `Exception` (doesn't have it as a declared or transitive superinterface). Which check is used by the lint determines whether `Never` is included.

I have no opinion on whether it should warn here. The outer `throw` is dead code, so that's reason enough to give *some* warning.

I'm not too worried about `foo();` not looking like it throws, because it should be named in a way which matches what it does, like `throwFailure` or just `fail`.
There is no unspoken convention that you shouldn't use `Never` returning functions to signify control flow.

**Comment by modulovalue** [comment]
> because it should be named in a way which matches what it does, like throwFailure or just fail.

I disagree with that. Everybody can name their identifiers in a way that implies that it will fail and they don't have to return Never, it can return anything even if it's named fail or exit or stop or quit. A naming convention does not enforce reliable guarantees that can be assumed to always hold.

The point of the broader suggestion (including #4054) is to help the user to transform their code into a "normal form" where certain assumptions can be assumed to be true in a sound way.

**Comment by modulovalue** [comment]
Your response seems to suggest that I might have been unclear in some aspects, I'm sorry if that is the case.

I am **not** proposing for the recommended style to be moved towards #4054. I merely want it **to be supported**. This issue is about `only_throw_errors` clashing with that proposal, and preventing people from quote: 

> You can obviously choose it for yourself,

doing exactly that.

---

To move this issue forward, would it be fine if `only_throw_errors` did not report on throwing a value that is know to be of type `Never`?

**Comment by lrhn** [comment]
I have no issue with `only_throw_errors` not complaining about throwing `Never`. That should solve your current problem.

However, I'd also have no problem with the analyzer deciding to give dead-code warning for the same code, which would probably not be a lint. If we ever choose to do that, you'd be back at needing `ignore` comments.
I don't want to promise that that won't happen. (Also won't promise that it definitely will.)

**Comment by modulovalue** [comment]
Thank you, Lasse.
